spmd/tensor/device_mesh.py

```diff
@@ -284,47 +284,72 @@ def scatter(
             scatter(tensor, scatter_list=None, src=src_for_dim, group=dim_group)
         return tensor
 
-    # pyre-fixme[3]: Return type must be annotated.
-    def broadcast(self, tensor: torch.Tensor, mesh_dim: int = 0):
+    def broadcast(self, tensor: torch.Tensor, mesh_dim: int = 0) -> torch.Tensor:
+        """
+        broadcast the tensor to a device mesh dimension. We by default
+        use the first rank of the mesh dimension as the source of truth, i.e
+        for a 2d mesh [[0, 1], [2, 3]], if we broadcast on mesh_dim = 1, we will
+        broadcast the tensor on rank 0 to rank 0/1, and tensor on rank 2
+        to rank 2/3.
+
+        Args:
+            tensor (torch.Tensor): tensor to scatter.
+            mesh_dim (int, optional): indicate which mesh dimension we want
+                to scatter on, we by default choose the first rank on the
+                mesh dimension as source of truth.
+
+        Returns:
+            A :class:`torch.Tensor` object
+        """
         dim_group = self._dim_groups[mesh_dim]
         # src need to be global rank
         src_for_dim = 0
         if dim_group is not GroupMember.WORLD:
             src_for_dim = get_global_rank(dim_group, 0)
 
-        return broadcast(tensor.contiguous(), src=src_for_dim, group=dim_group)
-
-    # pyre-fixme[3]: Return type must be annotated.
-    def all_gather(self, tensor: torch.Tensor, mesh_dim: int = 0):
-        dim_group = self._dim_groups[mesh_dim]
-        return all_gather(tensor, group=dim_group)
+        tensor = tensor.contiguous().clone()
```

Comment:
does clone trigger an additional copy, or is this something that happened anyways within the functional API beforehand?  (Also, if contigous actually triggered a copy already, then we don't need an additional copy?)

Reply:
Yes updated, thanks for the suggestion!